refactor: use VHD header to fetch block size

Millefeuille42 opened the pull request.

Millefeuille42 committed:
refactor: use VHD header to fetch block size

Signed-off-by: Mathieu Labourier <mathieu.labourier@vates.tech>

Millefeuille42 committed:
fix: decode ret if necessary

Signed-off-by: Mathieu Labourier <mathieu.labourier@vates.tech>

Millefeuille42 committed:
fix: fallback to VHD_BLOCK_SIZE if error in fetch

Signed-off-by: Mathieu Labourier <mathieu.labourier@vates.tech>

Millefeuille42 committed:
fix(test_vhdutil): mock getBlockSize

Signed-off-by: Mathieu Labourier <mathieu.labourier@vates.tech>

Millefeuille42 committed:
fix(test_FileSR): specify mock VHD_BLOCK_SIZE

Signed-off-by: Mathieu Labourier <mathieu.labourier@vates.tech>

Nambrok commented on libs/sm/cleanup.py:
```diff
@@ -1075,14 +1082,16 @@ def _getCoalescedSizeData(self):
         numBlocks = Util.countBits(blocksChild, blocksParent)
         Util.log("Num combined blocks = %d" % numBlocks)
-        sizeData = numBlocks * vhdutil.VHD_BLOCK_SIZE
+        sizeData = numBlocks * self.block_size
```
```suggestion
        sizeData = numBlocks * self.block_size()
```

Millefeuille42 replied:
`self.block_size` should be treated as an attribute due to the use of the `@property` decorator. I don't think your suggestion would work in that case 🤔

Wescoeur commented on libs/sm/cleanup.py:
```diff
@@ -2775,7 +2787,8 @@ def _finishCoalesceLeaf(self, parent):
     def _calcExtraSpaceNeeded(self, child, parent):
-        return lvhdutil.calcSizeVHDLV(parent.sizeVirt) - parent.sizeLV
+        return lvhdutil.calcSizeVHDLV(parent.sizeVirt, parent.block_size) - \
+            parent.sizeLV
```
Do you have a warn regarding line length? :)

Millefeuille42 replied:
I might have had one at a moment but forgot to readjust... Fixed in [fc386a4](https://github.com/xcp-ng/sm/pull/94/commits/fc386a416017a54daf4b7a10bc2d71ca35023776)

Wescoeur commented on libs/sm/drivers/FileSR.py:
```diff
@@ -561,7 +561,10 @@ def create(self, sr_uuid, vdi_uuid, size):
-                size = vhdutil.validate_and_round_vhd_size(int(size))
+                size = vhdutil.validate_and_round_vhd_size(
+                    int(size),
+                    vhdutil.VHD_BLOCK_SIZE
```
I think we need to rename it to `DEFAULT_VHD_BLOCK_SIZE`.

Millefeuille42 replied:
done in [0ea97f5](https://github.com/xcp-ng/sm/pull/94/commits/0ea97f563915a6ce8c173e9925fd754ff20b2212)

Wescoeur commented on libs/sm/vhdutil.py:
```diff
@@ -93,10 +93,29 @@ def ioretry(cmd, text=True):
+        ret = ioretry(cmd)
+    except util.CommandException as e:
+        util.SMlog("WARN: unable to fetch block size: {}".format(e))
+        return VHD_BLOCK_SIZE
```
I'm not sure that's safe to proceed like that.
Why couldn't we fetch this information?

Millefeuille42 replied:
This is to prevent cases where the path is wrong, or that the vhd-util binary could not be found. But now that you say it, it should be safer to not catch the exception indeed.

Millefeuille42 replied:
fixed in [4c866c7](https://github.com/xcp-ng/sm/pull/94/commits/4c866c7dd0f862c3404b24ec2c2361852aa4b908)

Wescoeur commented on tests/test_FileSR.py:
```diff
@@ -327,6 +327,7 @@ def test_create_vdi_vhd(self, mock_vhdutil):
         vdi = FakeFileVDI(sr, vdi_uuid)
         vdi.vdi_type = vhdutil.VDI_TYPE_VHD
         mock_vhdutil.validate_and_round_vhd_size.side_effect = vhdutil.validate_and_round_vhd_size
+        mock_vhdutil.VHD_BLOCK_SIZE = 2 * 1024 * 1024
```
```suggestion
        mock_vhdutil.DEFAULT_VHD_BLOCK_SIZE = vhdutil.DEFAULT_VHD_BLOCK_SIZE
```

Millefeuille42 replied:
fixed in [0ea97f5](https://github.com/xcp-ng/sm/pull/94/commits/0ea97f563915a6ce8c173e9925fd754ff20b2212)

Wescoeur commented on libs/sm/vhdutil.py:
```diff
@@ -93,10 +93,29 @@ def ioretry(cmd, text=True):
+    for field in fields:
+        field = field.strip()
+        if not field.startswith("Block size"): continue
+        return int(field.split(':')[1].strip().split(' ')[0])
```
```suggestion
        return int(field.split(':')[1].lstrip().split()[0])
```

Millefeuille42 replied:
done in [4c866c7](https://github.com/xcp-ng/sm/pull/94/commits/4c866c7dd0f862c3404b24ec2c2361852aa4b908)

Wescoeur commented on libs/sm/vhdutil.py:
```diff
@@ -93,10 +93,29 @@ def ioretry(cmd, text=True):
+        field = field.strip()
+        if not field.startswith("Block size"): continue
+        return int(field.split(':')[1].strip().split(' ')[0])
+    return VHD_BLOCK_SIZE
```
Should not happen. I suggest an exception instead.

Millefeuille42 replied:
done in [4c866c7](https://github.com/xcp-ng/sm/pull/94/commits/4c866c7dd0f862c3404b24ec2c2361852aa4b908)

Wescoeur commented on libs/sm/vhdutil.py:
```diff
@@ -93,10 +93,29 @@ def ioretry(cmd, text=True):
+        )
+    fields = ret.strip().split('\n')
+    for field in fields:
+        field = field.strip()
```
```suggestion
    for field in ret.split('\n'):
        field = field.lstrip()
```

Millefeuille42 replied:
fixed in [4c866c7](https://github.com/xcp-ng/sm/pull/94/commits/4c866c7dd0f862c3404b24ec2c2361852aa4b908)

Millefeuille42 committed:
style(cleanup): remove unnecessary line break in _calcExtraSpaceNeeded

Signed-off-by: Mathieu Labourier <mathieu.labourier@vates.tech>

Millefeuille42 committed:
refactor: rename VHD_BLOCK_SIZE to DEFAULT_VHD_BLOCK_SIZE

Signed-off-by: Mathieu Labourier <mathieu.labourier@vates.tech>

Millefeuille42 committed:
fix: adjust block size parsing

fix: remove default value if error in fetch
fix: raise an error instead of return default value if block size is not found

Signed-off-by: Mathieu Labourier <mathieu.labourier@vates.tech>
Co-authored-by: Ronan Abhamon <ronan.abhamon@vates.tech>
Signed-off-by: Mathieu Labourier <mathieu.labourier@vates.tech>

Millefeuille42 committed:
fix(test_LVHDSR): mock getBlockSize

Signed-off-by: Mathieu Labourier <mathieu.labourier@vates.tech>

Millefeuille42 committed:
fix(vhdutil): remove bytes to str conversion

Signed-off-by: Mathieu Labourier <mathieu.labourier@vates.tech>

Millefeuille42 committed:
fix(test_vhdutil): mock getBlockSize in test_get_vhd_info_allocated_size

Signed-off-by: Mathieu Labourier <mathieu.labourier@vates.tech>

Millefeuille42 committed:
test(test_vhdutil): test getBlockSize

Signed-off-by: Mathieu Labourier <mathieu.labourier@vates.tech>

Millefeuille42 commented on libs/sm/vhdutil.py:
```diff
@@ -93,10 +93,29 @@ def ioretry(cmd, text=True):
+        ret = ret.decode(
+            encoding = locale.getpreferredencoding(False),
+            errors="replace"
+        )
```
Turns out, as you pointed out, the tests were returning a bytes value in the vhd-util mock which caused the error. Since `getBlockSize` is mocked in said test, this is not a problem anymore. I removed this in [5c489fe](https://github.com/xcp-ng/sm/pull/94/commits/5c489feb8f2bcfe44d6a2f6a1742e9d26748a28f)

Millefeuille42 committed:
fix: add check for raw in block_size

Signed-off-by: Mathieu Labourier <mathieu.labourier@vates.tech>
